torch/nn/gradients.py

```diff
@@ -0,0 +1,209 @@
+        out_backprop : output gradient tensor (minibatch x out_channels x oH x oW)
+        weight: filters tensor (out_channels x in_channels/groups x kH x kW)
+        input_size : Shape of the input gradient tensor
+        stride (int or tuple, optional): Stride of the convolution. Default: 1
+        padding (int or tuple, optional): Zero-padding added to both sides of the input. Default: 0
+        dilation (int or tuple, optional): Spacing between kernel elements. Default: 1
+        groups (int, optional): Number of blocked connections from input channels to output channels. Default: 1
+        bias: optional bias tensor (out_channels). Default: None
+
+    Examples::
+
+        >>> out_backprop = autograd.Variable(torch.randn(1,1,3,2))
+        >>> weight = autograd.Variable(torch.randn(1,1,1,2))
+        >>> F.grad.conv2d_input(out_backprop, weight, stride=1, input_size=(1,1,3,3))
+    """
+    stride = _pair(stride)
+    padding = _pair(padding)
+    dilation = _pair(dilation)
+    kernel_size = (weight.shape[2], weight.shape[3])
+
+    if input_size is None:
+        raise ValueError("grad.conv2d_input requires specifying an input_size")
+
+    input_gradient_padding = _output_padding(out_backprop, input_size, stride,
+                                             padding, kernel_size)
+
+    return torch._C._VariableFunctions.conv_transpose2d(
+        out_backprop, weight, bias, stride, padding, input_gradient_padding,
+        groups, dilation)
+
+
+def conv2d_weight(in_channels, out_channels, input, out_backprop, stride=1,
+                  padding=0, dilation=1, groups=1, bias=None):
+    r"""
+    Computes the gradient of conv2d with respect to the weight of the convolution.
+
+    Args:
+        in_channels (int): Number of channels in the input image
+        out_channels (int): Number of channels in the output gradient
+        input: input tensor of shape (minibatch x in_channels x iH x iW)
+        out_backprop : output gradient tensor (minibatch x out_channels x oH x oW)
+        stride (int or tuple, optional): Stride of the convolution. Default: 1
+        padding (int or tuple, optional): Zero-padding added to both sides of the input. Default: 0
+        dilation (int or tuple, optional): Spacing between kernel elements. Default: 1
+        groups (int, optional): Number of blocked connections from input channels to output channels. Default: 1
+        bias: optional bias tensor (out_channels). Default: None
+
+    Examples::
+
+        >>> input = torch.autograd.Variable(torch.randn(1,1,3,3))
+        >>> out_backprop = torch.autograd.Variable(torch.randn(1,1,3,2))
+        >>> F.grad.conv2d_weight(1,1, input, out_backprop)
+    """
+    stride = _pair(stride)
+    padding = _pair(padding)
+    dilation = _pair(dilation)
+    dims = out_backprop.shape
+    min_batch = input.shape[0]
+
+    out_backprop_ = out_backprop.repeat(1, in_channels // groups, 1, 1)
```

Comment:
The code here looks more complicated than I thought it should be. What's going on here?

Reply:
For calculating the gradient with respect to weight, the `out_backprop` is the weight for the `conv2d` and the weight gradient is the output.  The `out_backprop` is expanded and reshaped so that the forward convolution results in the (weight gradient) having the same shape as the `weight`. The `out_backprop` is expanded by the number of input channels divided by the number of groups (`in_channels // groups`). Then the `input` is reshaped so that it has channels and batches merged and the `conv2d` forward pass with groups = `in_channels * min_batch`. 

The result is reshaped with dimension 0 as the number of batches and the tensor is summed along that dimension. After that, the result(`grad_w`) is reshaped and narrowed to fit the `weight_size`.